src/kernels/raw/finder/localKernelSpecFinderBase.node.ts

```diff
@@ -125,7 +133,64 @@ export abstract class LocalKernelSpecFinderBase<
         // ! as the has and set above verify that we have a return here
         return this.kernelSpecCache.get(cacheKey)!.promise;
     }
+    protected async listKernelsFirstTimeFromMemento(cacheKey: string): Promise<T[]> {
+        // Check memento too
+        const cache = this.globalState.get<{ kernels: T[]; extensionVersion: string }>(cacheKey, {
+            kernels: [],
+            extensionVersion: ''
+        });
+
+        let kernels: T[] = [];
+        /**
+         * The cached list of raw kernels is pointing to kernelSpec.json files in the extensions directory.
+         * Assume you have version 1 of extension installed.
+         * Now you update to version 2, at this point the cache still points to version 1 and the kernelSpec.json files are in the directory version 1.
+         * Those files in directory for version 1 could get deleted by VS Code at any point in time, as thats an old version of the extension and user has now installed version 2.
+         * Hence its wrong and buggy to use those files.
+         * To ensure we don't run into weird issues with the use of cached kernelSpec.json files, we ensure the cache is tied to each version of the extension.
+         */
+        if (cache && Array.isArray(cache.kernels) && cache.extensionVersion === this.env.extensionVersion) {
+            kernels = cache.kernels.map(deserializeKernelConnection) as T[];
+        }
+
+        // Validate
+        const validValues: T[] = [];
+        await Promise.all(
+            kernels.map(async (item) => {
+                if (await this.isValidCachedKernel(item)) {
+                    validValues.push(item);
+                }
+            })
+        );
+        return validValues;
+    }
+
+    protected async writeToMementoCache(values: T[], cacheKey: string) {
+        const serialized = values.map(serializeKernelConnection);
+        await Promise.all([
+            removeOldCachedItems(this.globalState),
```

Comment:
Just to note, previously the cache keys used to be defined in the commonFinder.ts file where removeOldCachedItems was defined. Might still want to define them there so that they can be removed when versions update? I would still take this PR and just file an issue for that as it shouldn't hurt anything now.